sopht/simulator/immersed_body/cosserat_rod/cosserat_rod_forcing_grids.py

```diff
@@ -321,20 +326,112 @@ def __init__(
+                    reference_length = max(
+                        rod_end_radius * grid_angular_spacing,
+                        np.amax(self.cosserat_rod.lengths),
+                    )
+                    # note: we include at least one point on the element for the cap
+                    end_elem_radial_grid_density = max(
+                        int(rod_end_radius // reference_length), 1
+                    )
+                else:
+                    # the rod surface grid is already the minimum anyway (single point)
+                    # no point adding more grid points for end surface
+                    end_elem_radial_grid_density = 0
+
+                end_elem_grid_points_radius_ratio.append(
+                    np.linspace(
+                        0, rod_end_radius, end_elem_radial_grid_density, endpoint=False
+                    )
+                    / rod_end_radius
+                )
+                # linearly scale the number of grid points towards outer most surface
+                end_elem_surface_grid_points.append(
+                    np.linspace(
+                        1,
+                        self.surface_grid_points[rod_end_idx],
+                        end_elem_radial_grid_density,
+                        endpoint=False,
+                    ).astype(int)
+                )
+
+                # update surface grid points at the end elements
+                self.surface_grid_points[rod_end_idx] += end_elem_surface_grid_points[
+                    rod_end_idx
+                ].sum()
+
+        # finalize number of lag nodes
         num_lag_nodes = self.surface_grid_points.sum()
         super().__init__(grid_dim, num_lag_nodes)
-        self.n_elems = cosserat_rod.n_elems
+
+        # store grid point radius ratio for each forcing point
+        self.grid_point_radius_ratio = np.ones((self.num_lag_nodes))
+        if self.with_cap:
+            # Update grid point radius ratio for relevant points (rod ends)
+            for rod_end_idx in [0, -1]:
+                # start index for grid points within the surface grid point of end element
+                # we store the inner grid points at the end portion of the array window
+                # corresponding to the element
+                start_idx = (
+                    self.surface_grid_points.cumsum()[rod_end_idx]
+                    - end_elem_surface_grid_points[rod_end_idx].sum()
+                )
+                # loop over each concentric circle and update the radius ratio
+                for idx, num_grid_points in enumerate(
+                    end_elem_surface_grid_points[rod_end_idx]
+                ):
+                    end_idx = start_idx + num_grid_points
+                    self.grid_point_radius_ratio[
+                        start_idx:end_idx
+                    ] = end_elem_grid_points_radius_ratio[rod_end_idx][idx]
+                    start_idx = end_idx
```

Comment:
I think all this makes the init a bit complicated... Maybe we should refactor it into a free function or a member function and then call it in init? That function can also have a minimal docstring that explains what's happening.

Reply:
Attempted fix in [a8ad220](https://github.com/SophT-Team/SophT/pull/121/commits/a8ad220cef0a4b6d88d60473cfe907e20f1b3766)